httpserver/service_node.cpp

```diff
@@ -26,6 +26,54 @@ using service_node::storage::Item;
 namespace loki {
 using http_server::connection_t;
 
+constexpr std::array<std::chrono::seconds, 5> RETRY_INTERVALS = {
+    std::chrono::seconds(5), std::chrono::seconds(10), std::chrono::seconds(20),
+    std::chrono::seconds(40), std::chrono::seconds(80)};
+
+FailedWork::FailedWork(boost::asio::io_context& ioc, const sn_record_t& sn,
+                       const std::string& msg_body)
+    : ioc_(ioc), retry_timer_(ioc), sn_(sn), msg_body_(msg_body) {}
+
+void FailedWork::retry(std::shared_ptr<FailedWork>&& self) {
+
+    attempt_count_ += 1;
+    if (attempt_count_ > RETRY_INTERVALS.size()) {
+        BOOST_LOG_TRIVIAL(debug)
+            << "Gave up after " << attempt_count_ << " attempts";
+        return;
+    }
+
+    retry_timer_.expires_after(RETRY_INTERVALS[attempt_count_ - 1]);
+    BOOST_LOG_TRIVIAL(debug)
+        << "Will retry in " << RETRY_INTERVALS[attempt_count_ - 1].count()
+        << " secs";
+
+    retry_timer_.async_wait(
+        [self = std::move(self)](const boost::system::error_code& ec) mutable {
+            request_t req;
+            req.body() = self->msg_body_;
+            req.target("/v1/swarms/push");
```

Comment:
Maybe there's a better way to include the retry mechanism.
Currently `FailedWork` seems to duplicate a lot of the behaviour of `make_http_request` with the addition of counting the number of failures. The duplication of `req.target("/v1/swarms/push");` is a bit of a red flag.

Since this is all asynchronous and objects live (or survive) by passing shared pointers around to callbacks, maybe a new object is needed to store the request state (number of attempts etc) and pass that around through the normal code flow. Not sure how to express that, let me know if that makes sense?

Reply:
I know this would come up, and I gave it some thought earlier. I could reduce code duplication by constructing `FailedWork` directly in `relay_one` as "attempt 0", but that would mean a lot more overhead (creating timers etc) for the much more common case of successful push...

Reply:
But I guess I could construct `request_t` once and store in inside `FailedWork`.

Reply:
Now constructing request_t only once!